### PR TITLE
feat(GitHub Lines): parse and send the content of github urls

### PR DESCRIPTION
**Pull request by JPBM135**
## What this does?
This bit of code will detect and parse any GitHub URLs that contain specific lines and present them in chat!

### **Features**
- [x] Multiple urls
- [x] Full files

### **Type**
- [x] Normal Github urls
- [ ] ~~Diff files~~ (Will be added in another PR)
- [x] Gists

### **~~Bugs~~ Features**
- [x] File size limit
- [x] Reverse numbers if people are dumb (L2-L1 => L1-L2)
- [x] Convert 0 to 1
- [x] Limit line size to the file size
- [x] Hide embed if applicable

<details>
<summary>Examples</summary>

### Examples:
Thread opening for normal text channels:
image

Limited text on channels that does not support threads:
image

</details>
<details>
<summary>Test suite</summary>

## Test Suite

Test 1: Full file
```
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/migrations/1625779685-next.js>
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/migrations/1625779685-next.js#La>
```

Test 2: One Line
```
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/migrations/1625779685-next.js#L10>
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/migrations/1625779685-next.js#L50>
```

Test 3: Range
```
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/migrations/1625779685-next.js#L10-L50>
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/migrations/1625779685-next.js#L30-L50>
```

Test 4: Gist
```
<https://gist.github.com/almostSouji/71a33e5f8e84a0960b4ed3a1609915f5#file-cookiecollections-md>
<https://gist.github.com/almostSouji/71a33e5f8e84a0960b4ed3a1609915f5#file-cookiecollections-md-L13>
<https://gist.github.com/almostSouji/71a33e5f8e84a0960b4ed3a1609915f5#file-cookiecollections-md-L13-L15>
<https://gist.github.com/JPBM135/4e8c72128a3f0fd28ffc41a983c1bdb2#file-testing-js-L13>
```

Test 5: Multiple files (send in one message)
Test 5.1: Threads, should cap the urls
```
<https://github.com/JPBM135/yuudachi/blob/main/.vscode/settings.json#L1-L10>
<https://github.com/JPBM135/yuudachi/blob/main/.vscode/settings.json#L11-L14>
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/migrations/1625779685-next.js#L10-L50>
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/.lintstagedrc.js#L1>
<https://github.com/JPBM135/yuudachi/blob/main/.vscode/settings.json>
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/migrations/1625779685-next.js#L100-L142>
```

Test 6: Should remove trailing tabs:
```
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/migrations/1625779685-next.js#L4-L13>
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/migrations/1625779685-next.js#L20-L66>
<https://github.com/discordjs/guide/blob/main/guide/.vuepress/styles/sidebar.scss#L9-L19>
<https://github.com/discordjs/discord.js/blob/main/packages/discord.js/src/client/Client.js#L43-L64>
```

Test 7: Shouldn't remove trailing tabs:
```
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/migrations/1625779685-next.js#L2-L19>
<https://github.com/discordjs/discord.js/blob/main/package.json#L1-L15>
<https://github.com/discordjs/discord.js/blob/main/packages/discord.js/src/client/Client.js#L42-L64>
```

Test 8: Special cases (dumbs dumbs)
- File bigger than 8Mb (shouldn't output)
```
<https://github.com/kerrickstaley/Chinese-Vocab-List/blob/master/reference_files/subtlex_ch.tsv>
```
- Reverse
```
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/migrations/1625779685-next.js#L50-L10>
```
- Really long path (obv an Java app)
```
<https://github.com/LorittaBot/Loritta/blob/main/discord/loritta-bot-common/src/main/kotlin/net/perfectdreams/loritta/cinnamon/discord/gateway/modules/StarboardModule.kt#L1-L54>
```
- L = 0
```
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/migrations/1625779685-next.js#L0>
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/migrations/1625779685-next.js#L0-L10>
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/migrations/1625779685-next.js#L50-L0>
```
- Infinity
```
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/migrations/1625779685-next.js#L100000000>
<https://github.com/JPBM135/yuudachi/blob/main/packages/yuudachi/migrations/1625779685-next.js#L0-L100000000>
```

</details>

### CODE REVIEW

**Comment by jaw0r3k** [comment]
but why not as slash command?
> as message it will be kindof spam

File: src/events/gitHub/messageCreate.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import { logger } from "@yuudachi/framework";
+import type { Event } from "@yuudachi/framework/types";
+import { type Message, Events, Client } from "discord.js";
+import { injectable } from "tsyringe";
+import { handleGithubUrls } from "../../functions/gitHub/handler.js";
+import { ASSISTCHANNELS } from "../../util/constants.js";
+
+@injectable()
+export default class implements Event {
+	public name = "Message GitHub lines resolvable";
```
Comment (almostSouji) [inline src/events/gitHub/messageCreate.ts]:
```suggestion
	public name = "Resolve GitHub links";
```

File: src/util/gitHub.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export function convertUrlToRawUrl(url: string) {
+	return url.replace("github.com", "raw.githubusercontent.com").replace("/blob", "");
+}
```
Comment (almostSouji) [inline src/util/gitHub.ts]: newline EOF (ci red)
as it is directly related to the github command, can be placed in that file, it's not a general utility with many use cases

File: src/functions/gitHub/handler.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
+import type { VoiceChannel } from "discord.js";
```
Comment (almostSouji) [inline src/functions/gitHub/handler.ts]: should be `functions/github/handler.ts`

File: src/util/ellipsis.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+export function ellipsis(text: string, maxLength: number) {
```
Comment (almostSouji) [inline src/util/ellipsis.ts]: import from framework instead

**Review by almostSouji — CHANGES_REQUESTED** [review]

File: src/functions/gitHub/handler.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+		const url = convertUrlToRawUrl(match[0]!);
+
+		const rewFile = await request(url).then(async (res) => res.body.text());
+		const linesRequested = rewFile.split("\n").slice(startLine - 1, end ? endLine : startLine);
+
+		const lang = path!.split(".").pop() ?? "ansi";
+
+		const content = [
+			`${end ? `Lines ${inlineCode(start!)} to ${inlineCode(end)}` : `Line ${inlineCode(start!)}`} of ${italic(
+				`${org}/${repo}${path}`,
```
Comment (almostSouji) [inline src/functions/gitHub/handler.ts]: - Should state the actual lines it displays (10 to 50, if limited to 10, makes no sense. Should be 10 to 20)
- Any specific reason why this isn't a link?

File: src/functions/gitHub/handler.ts [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+						.join("\n"),
+					maxLength,
+				),
+			),
+		);
+
+		if (!thread) {
+			thread = await message.startThread({
+				name: `GitHub Lines for this message`,
+				reason: "Message contained a GitHub URL that points to specific lines",
```
Comment (almostSouji) [inline src/functions/gitHub/handler.ts]:
```suggestion
				reason: "Resolving GitHub link",
```

**Review by almostSouji — CHANGES_REQUESTED** [review]

File: src/events/gitHub/messageCreate.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
+import { on } from "node:events";
+import { setTimeout as wait } from "node:timers/promises";
+import { logger } from "@yuudachi/framework";
+import type { Event } from "@yuudachi/framework/types";
+import { type Message, Events, Client } from "discord.js";
+import { injectable } from "tsyringe";
+import { handleGithubUrls } from "../../functions/gitHub/handler.js";
+import { ASSISTCHANNELS } from "../../util/constants.js";
```
Comment (almostSouji) [inline src/events/gitHub/messageCreate.ts]: `ASSISTCHANNELS` is only support forums, i'm not sure why you depend it to that?

**Review by almostSouji — CHANGES_REQUESTED** [review]

**Comment by almostSouji** [comment]
> but why not as slash command?
> as message it will be kindof spam

~~I don't see it very useful as slash command because people are not necessarily aware of it existing~~
~~what I could imagine is a message context menu to prompt the thread/resolving instead of doing it automatically~~

after some discussion we will keep it automated for now and see how that goes

File: src/functions/gitHub/handler.ts (hunk 5 above)
Comment (JPBM135) [reply src/functions/gitHub/handler.ts]: bullet one: working on that
bullet two: the original message is already a link, I don't see the need of being a link to, but can be tbh

File: src/events/gitHub/messageCreate.ts (hunk 7 above)
Comment (JPBM135) [reply src/events/gitHub/messageCreate.ts]: Delay for assist channels is increased so the files will sent after the autoposter

File: src/functions/gitHub/handler.ts (hunk 5 above)
Comment (JPBM135) [reply src/functions/gitHub/handler.ts]: Bullet one: Fixed in 5dcfa16

File: src/util/ellipsis.ts (hunk 4 above)
Comment (JPBM135) [reply src/util/ellipsis.ts]: Fixed in 277127c

File: src/functions/gitHub/handler.ts (hunk 3 above)
Comment (JPBM135) [reply src/functions/gitHub/handler.ts]: Fixed in 277127c

File: src/util/gitHub.ts (hunk 2 above)
Comment (JPBM135) [reply src/util/gitHub.ts]: Fixed in 277127c

File: src/functions/github/handler.ts [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+	codeBlock,
+	inlineCode,
+	italic,
+	ChannelType,
+} from "discord.js";
+import kleur from "kleur";
+import { request } from "undici";
+
+// eslint-disable-next-line unicorn/no-unsafe-regex
+const GitHubUrlRegex = /https:\/\/github.com\/(?<org>.+?)\/(?<repo>.+?)\/blob(?<path>\/[^\n#>]+)(?<opts>.+)?/g;
```
Comment (MrMythicalYT) [inline src/functions/github/handler.ts]:
```suggestion
const GitHubUrlRegex = /https:\/\/github\.com\/(?<org>.+?)\/(?<repo>.+?)\/blob(?<path>\/[^\n#>]+)(?<opts>.+)?/g;
```
Escape the `.` in the regex

**Comment by jaw0r3k** [comment]
Also next thing 
What about file limit when sending fullFile?
> maybe little hard to achieve but possible

File: src/functions/github/handler.ts (hunk 8 above)
Comment (JPBM135) [reply src/functions/github/handler.ts]: Thanks, fixed

**Comment by JPBM135** [comment]
moved to opening post in a toggle ~Souji

File: src/util/array.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
+export function stringArrayLength(arr: string[]) {
+	return arr.reduce((acc, cur) => acc + cur.length, 0);
+}
+
+export function arrayEllipsis(arr: string[], maxLength: number) {
```
Comment (JPBM135) [inline src/util/array.ts]: Rename
```suggestion
export function truncateArrayJoin(arr: string[], maxLength: number): string[] {
```

File: src/util/constants.ts [hunk 10]
```diff
@@ -20,3 +20,6 @@ export const TAB = "\u200B \u200B \u200B" as const;
 
 export const EMOJI_NEWBIE = "<:newbie:962332319623049226>" as const;
 export const ASSISTCHANNELS = ["986520997006032896", "998942774994927646"];
+
+export const URL_REGEX =
+	/(?<protocol>https?:\/\/(?:www\.|(?!www))[\dA-Za-z][\dA-Za-z-]+[\dA-Za-z]\.\S{2,}|www\.[\dA-Za-z][\dA-Za-z-]+[\dA-Za-z]\.\S{2,}|https?:\/\/(?:www\.|(?!www))[\dA-Za-z]+\.\S{2,}|www\.[\dA-Za-z]+\.\S{2,})/g;
```
Comment (JPBM135) [inline src/util/constants.ts]: This capture group is not the protocol!

```suggestion
	/(?<url>https?:\/\/(?:www\.|(?!www))[\dA-Za-z][\dA-Za-z-]+[\dA-Za-z]\.\S{2,}|www\.[\dA-Za-z][\dA-Za-z-]+[\dA-Za-z]\.\S{2,}|https?:\/\/(?:www\.|(?!www))[\dA-Za-z]+\.\S{2,}|www\.[\dA-Za-z]+\.\S{2,})/g;
```

File: src/functions/github/utils.ts [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+	fullFile: boolean;
+	onThread: boolean;
+	path: string;
+	startLine: number;
+};
+
+export function convertUrlToRawUrl(url: string) {
+	return (
+		url
+			// eslint-disable-next-line prefer-named-capture-group
```
Comment (JPBM135) [inline src/functions/github/utils.ts]: I think this can be removed

```suggestion

```

File: src/functions/github/handler.ts [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+
+	const matches = Array.from(urls).map(([url]) => {
+		const match = validators.find((validator) => validator.regex.exec(url!));
+		if (!match) return null;
+
+		return {
+			regexMatch: new RegExp(match.regex, "").exec(url!),
+			...match,
+		};
+	});
```
Comment (JPBM135) [inline src/functions/github/handler.ts]: Instead of returning on the loop, filter here

File: src/functions/github/handler.ts [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+export async function handleGithubUrls(message: Message<true>) {
+	const urls = new Set(message.content.matchAll(URL_REGEX));
+	if (!urls.size) return;
+
+	const matches = Array.from(urls).map(([url]) => {
+		const match = validators.find((validator) => validator.regex.exec(url!));
+		if (!match) return null;
+
+		return {
+			regexMatch: new RegExp(match.regex, "").exec(url!),
```
Comment (JPBM135) [inline src/functions/github/handler.ts]: Return null if the regex match doesn't match, also the regex isn't global anymore
```suggestion
		if (!match) return null;

		const regexMatch = match.regex.exec(url!);
		if (!regexMatch) return null;

		return {
			regexMatch,
```

File: src/functions/github/handler.ts [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+			...match,
+		};
+	});
+
```
Comment (JPBM135) [inline src/functions/github/handler.ts:86]: Add a check to see if the matches have any length

File: src/functions/github/handler.ts [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+		"GitHub URL(s) detected",
+	);
+
+	let thread: AnyThreadChannel<boolean> | VoiceChannel | null = isOnThread
+		? (message.channel as AnyThreadChannel<boolean> | VoiceChannel)
+		: null;
+
+	const tempContent: string[] = [];
+
+	for (const match of matches) {
```
Comment (JPBM135) [inline src/functions/github/handler.ts]: Refactor this loop to be less wild

**Review by almostSouji — APPROVED** [review]

**Pull request merged**